Compilation database support - clang, gcc, mingw

RevoluPowered opened the pull request:
This file is from mongodb and the tooling Automatically generates the compilation database across all platforms.

Thanks to acmarrow for a simple solution

https://github.com/SCons/scons/issues/3355

For the uninitiated this supports IDE symbol lookup, refactoring tools, documentation tools:
This means we can support intelisense style technologies with reliable type lookup.

For example:
- CLion symbols (https://www.jetbrains.com/help/clion/compilation-database.html)
- VSCode symbols / refactoring (https://devblogs.microsoft.com/cppblog/visual-studio-code-cc-extension-october-2017-update/)
- QTcreator (https://doc.qt.io/qtcreator/creator-clang-codemodel.html#using-compilation-databases)
- atom (https://atom.io/packages/linter-clang)
- neovim
- vim https://github.com/ycm-core/YouCompleteMe
- anything which supports compilation db
- etc

Known issue:
- no support for msvc compiler

Co-authored-by: K. S. Ernest (iFire) Lee <ernest.lee@chibifire.com>

RevoluPowered commented:
Python 2.7 justification for upgrade:
- Python 2.7 is going to be depreciated very soon in scons. (next release)

Scons would not build without this update with the compilation db on windows, due to a bug they have not fixed.

Simplifies osx path code.

Updated platform install to use sudo to install scons to prevent us having to modify path. 

Since the VM image is thrown away this doesn't matter.

gavinws commented on .travis.yml:
```diff
@@ -7,6 +7,9 @@ stages:
   - build
 
+python:
+  - "3.7"
```
This doesn't do what you intended, as you can see in the build logs Python 2.7 is still being used. Instead you will need to do one of the following:

* during `install` phase add command `pyenv global system 3.7` (probably not best option IMO)
* add a `.python-version` file with `3.7` as contents to the root of this repo
* add to `env` the `PYENV_VERSION` variable with `3.7` value

You may also need to use `pip3` instead of `pip` to install scons. Also, unless you use the first option, `sudo` should be unnecessary.

RevoluPowered replied:
The correction can be simpler thanks for pointing this out, swapping out pip for pip3 will mean it will use a pip3 variant of scons, which has python3 shebang.

RevoluPowered replied:
After re-reading documentation you're right we would need the python specifier and lang set to python.

I think there might be an easier way will do a bit of research

IAmActuallyCthulhu commented:
I have tested it on macOS (10.14.6) with CLion (2019.2.4).

While testing I found one issue, Objective-C files found in [/platform/osx](https://github.com/godotengine/godot/tree/master/platform/osx) are **not** recorded to the compilation database. 

![CLion error message](https://user-images.githubusercontent.com/24380491/67165061-60467000-f346-11e9-98ab-52bca3d9510d.png)

I was able to fix it (quick and dirty) by adding `, '.mm'` to the `_CXXSuffixes` array.

https://github.com/godotengine/godot/blob/fb8bb2dcd4db1ece6bb25aabd52837b2f363b984/site_scons/site_tools/compilation_db.py#L39

---

P.S. Thank you @RevoluPowered and @fire for making this PR. I am incredibly grateful, happy, and overall ecstatic to have this.

IAmActuallyCthulhu commented:
I have tested it on macOS (10.14.6) with VSCode.

VSCode detected the `compile_commands.json` and generated the config `c_cpp_properties.json`. There are no false positives anymore!

However, same Objective-C issues as explained in https://github.com/godotengine/godot/pull/32848#issuecomment-544287242.

![image](https://user-images.githubusercontent.com/24380491/67166768-0fd80e00-f358-11e9-8475-d91ad3e136d1.png)

Side Note: I do not know if the generated `c_cpp_properties.json` file is correct or needs manual changes.

RevoluPowered commented:
Hello, I will update the PR tomorrow as we are all returning from GodotCon

Calinou commented on .travis.yml:
```diff
@@ -2,6 +2,7 @@ language: cpp
 
 # OS config, depends on actual 'os' in build matrix
 dist: xenial
+sudo: true
```
`sudo: true` is deprecated since December 2018, see https://changelog.travis-ci.com/the-container-based-build-environment-is-fully-deprecated-84517

RevoluPowered replied:
Thanks! Removed this :)

Calinou commented on .travis.yml:
```diff
@@ -108,19 +109,22 @@ matrix:
+  - sudo ln -s /usr/bin/python3 /usr/bin/python
+  - sudo apt-get update
+  - sudo apt-get install -y python3 python3-pip python3-setuptools
+  - sudo python -m pip install scons
```
I'd look at using the `--user` pip switch then appending `$HOME/.local/bin` to the `PATH` so you can avoid `sudo pip` and all its caveats :slightly_smiling_face:

```shell
python -m pip install --user scons
export PATH="$HOME/.local/bin:$PATH"
scons …
```

RevoluPowered replied:
Thanks! I have updated the code to use this method as it's cleaner.

RevoluPowered commented:
Note: @akien-mga please do not merge this yet travis is not building this PR anymore and I am not sure why.

RevoluPowered commented:
not got access to apple hw, will delay this until tomorrow

RevoluPowered committed:
Added compilation database support for clang and gcc

This tool is originally from mongodb.

- Updated CPPSUFFIXES to use scons suffixes
- objective-c files will also be loaded into the compilation database where the compiler / tooling is available to compile the files.

Known limitations:

- This will not work with msvc as your compiler.

RevoluPowered commented:
@IAmActuallyCthulhu updated the suffixes properly so should be platform agnostic now

Adding the method to the cxxsuffix as requested would have broken it, so I have made it retrieve this directly from scons, on osx it will make it available and linux, but on windows it will not as this made appveyor break.

Xrayez commented:
Doesn't seem to work on Windows, VS2017, Python 3.6.4, Scons v3.0.5.
```
[ 96%] Building compilation database compile_commands.json
...
[100%] scons: done building targets.
```
It simply produces an empty array in `compile_commands.json`.

```
scons platform=windows tools=yes target=release_debug bits=64 -j8
```

Same thing with `target=debug debug_symbols=yes`.

RevoluPowered commented:
> Doesn't seem to work on Windows, VS2017, Python 3.6.4, Scons v3.0.5.
> 
> ```
> [ 96%] Building compilation database compile_commands.json
> ...
> [100%] scons: done building targets.
> ```
> 
> It simply produces an empty array in `compile_commands.json`.
> 
> ```
> scons platform=windows tools=yes target=release_debug bits=64 -j8
> ```
> 
> Same thing with `target=debug debug_symbols=yes`.

This will only work with scons 3.1.1 the stable release

RevoluPowered commented:
vscode currently requires us to reload this compilation_db file manually, I think I can update the behaviour so that vscode just sees this as a file modification rather than the file being unreadable during write.

Thorhian commented:
Hi RevoluPowered! I also wanted to suggest adding "compile_commands.json" to the .gitignore file for the project as well. That wouldn't cause any unwanted side effects, would it?

RevoluPowered commented:
> Hi RevoluPowered! I also wanted to suggest adding "compile_commands.json" to the .gitignore file for the project as well. That wouldn't cause any unwanted side effects, would it?

Good suggestion I will test this and see what happens

clayjohn commented:
I have the same issue as Xrayez. I am using Windows 10 and I have scons version 3.1.1 installed. The compilation database just contains an empty array.

RevoluPowered commented:
> 
> 
> I have the same issue as Xrayez. I am using Windows 10 and I have scons version 3.1.1 installed. The compilation database just contains an empty array.

ah okay will look into this later tonight thanks both!

RevoluPowered commented:
Rebased on master

I can reproduce this issue on windows 10, but only when I use msvc compiler, I was targeting the gcc and mingw.

Seems the issue is msvc only bug, and the file is indeed empty []. 

Initial research shows modifications I made to CXX_SUFFIX is working on this platform, so will need to do more looking tomorrow.

akien-mga commented on site_scons/site_tools/compilation_db.py:
```diff
@@ -0,0 +1,173 @@
+# Copyright 2015 MongoDB Inc.
+#
```
Why the extra `site_scons/site_tools` path?

RevoluPowered replied:
scons loads these before sconstruct files, it's annoying to have to have the extra directory but I do not know how to make this work without the extra path

here is some docs on this:
https://scons.org/doc/1.0.1/HTML/scons-user/x3627.html

akien-mga replied:
That sucks. I'm pretty territorial about what folders we have in our Git repo, so I really don't want to add something like this as top-level folder.
Isn't there a way to declare a new `Tool` from an imported module (`import compilation_db`)?

Otherwise upstream seems willing to merge this upstream, so that might be the cleanest way forward :)

RevoluPowered replied:
I think the best solution would be to merge with scons directly, I like that idea better and means less code here

akien-mga commented on .travis.yml:
```diff
@@ -103,21 +113,30 @@ matrix:
-  - if [ "$TRAVIS_OS_NAME" = "osx" ]; then
-      export PATH=${PATH}:/Users/travis/Library/Python/2.7/bin;
+  - if [ "$TRAVIS_OS_NAME" = "osx" ] && [ "$PLATFORM" = "android" ]; then
+      misc/travis/android-tools-osx.sh;
```
We're no longer building Android on OSX, and this script has been removed a while ago.

RevoluPowered replied:
okay will rebase and remove

akien-mga commented on .travis.yml:
```diff
@@ -103,21 +113,30 @@ matrix:
+  - source buildenv/bin/activate
+  - python3 -m pip install --upgrade pip
+  - pip install scons
+  - scons --version
```
This can be simplified a lot by using the pre-existing Python 3 and pip of the Xenial environment. I experimented a bit, I'll make a separate PR to upgrade to Python 3.

akien-mga replied:
I pushed the changes as #36074, so if you rebase you should be able to drop all `.travis.yml` changes.

RevoluPowered replied:
thank you!, I will rebase without the edits to those files

RevoluPowered commented:
@akien-mga I have updated with those changes requested and squashed into a single commit

RevoluPowered commented:
Just leaving a note here before closing: 
- we are going to attempt getting this merged with scons.

This tooling is compatible easily with 4.0 if you need a solution (now) for compilation databases this is still the most efficient method of producing the compilation db. it seriously outperforms bear, and intercept-build as the list is already stored in scons data.

We will hopefully get support for the msvc compiler, this may require upstream fixes in scons to get right.

So hopefully in a month or two you will recieve a scons update with this functionality built into it.

clayjohn commented:
Thanks for all your work :)

I plan on running MinGW periodically to update my compilation database in the meantime.

RevoluPowered commented:
> Thanks for all your work :)
> 
> I plan on running MinGW periodically to update my compilation database in the meantime.

Ach, can only take a minor part of credit, was team effort between me and @fire :) 

also, acmarrow from mongodb :) and mongodb themselves

akien-mga commented:
Force pushed an amend that fixes Python style issues (we now use `black` for formatting), and moves the script to `misc/scons/site_tools` as I found a workaround to override the default `site_scons` path. (Basically reimplementing what SCons' `--site-dir` command line option would do by using a private method - Thanks Python for not having real private methods :P).

akien-mga commented:
Thanks!

RevoluPowered commented:
Thanks for your help merging I had little time to pull away from FBX. That's great I am so happy, I use this regularly! <3

Faless commented:
Getting the following error after this merge:
```
[ 98%] scons: done building targets.
TypeError: changed_since_last_build_node() missing 1 required positional argument: 'node':
  File "/usr/lib/scons/SCons/Script/Main.py", line 1376:
    _exec_main(parser, values)
  File "/usr/lib/scons/SCons/Script/Main.py", line 1339:
    _main(parser)
  File "/usr/lib/scons/SCons/Script/Main.py", line 1103:
    nodes = _build_targets(fs, options, targets, target_top)
  File "/usr/lib/scons/SCons/Script/Main.py", line 1313:
    jobs.run(postfunc = jobs_postfunc)
  File "/usr/lib/scons/SCons/Job.py", line 111:
    self.job.start()
  File "/usr/lib/scons/SCons/Job.py", line 423:
    task.executed()
  File "/usr/lib/scons/SCons/Script/Main.py", line 237:
    SCons.Taskmaster.OutOfDateTask.executed(self)
  File "/usr/lib/scons/SCons/Taskmaster.py", line 317:
    t.release_target_info()
  File "/usr/lib/scons/SCons/Node/FS.py", line 2980:
    self.changed(allowcache=True)
  File "/usr/lib/scons/SCons/Node/FS.py", line 3223:
    has_changed = SCons.Node.Node.changed(self, node)
  File "/usr/lib/scons/SCons/Node/__init__.py", line 1455:
    if _decider_map[child.changed_since_last_build](child, self, prev_ni):
```

Tested with both pip version:
```
SCons by Steven Knight et al.:
	script: v3.1.2.bee7caf9defd6e108fc2998a2520ddb36a967691, 2019-12-17 02:07:09, by bdeegan on octodog
	engine: v3.1.2.bee7caf9defd6e108fc2998a2520ddb36a967691, 2019-12-17 02:07:09, by bdeegan on octodog
```

and distro version:
```
SCons by Steven Knight et al.:
	script: v3.0.1.74b2c53bc42290e911b334a6b44f187da698a668, 2017/11/14 13:16:53, by bdbaddog on hpmicrodog
	engine: v3.0.1.74b2c53bc42290e911b334a6b44f187da698a668, 2017/11/14 13:16:53, by bdbaddog on hpmicrodog

```

akien-mga commented:
@Faless The backtrace is from your distro version (`/usr/lib/scons`), so make sure that the pip version is not using the wrong SCons lib from the distro.

akien-mga commented:
We should likely increase `EnsureSConsVersion` to the version needed for the compile DB, or alternatively disable generation of the compile DB on older SCons versions (might be more user-friendly).

Faless commented:
@akien-mga right, my bash alias tricked me -.- . It works with the pip version, so yeah, `scons 3.0.1` doesn't seem to work, `scons 3.1.2` is fine.

akien-mga commented:
Cherry-picked for 3.2.2. (Together with #38830 to fix build on older SCons versions.)
